### PR TITLE
kube-proxy ipvs: fix to prevent concurrent map read and map write

### LINKED ISSUE #103260: ipvs: Concurrent map read and map write exist in kube-proxy
#### What happened:
image

#### What you expected to happen:
no panic

#### Anything else we need to know?:
https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/proxy/ipvs/graceful_termination.go#L86

#### Environment:
- Kubernetes version (use `kubectl version`): k8s 1.16

### PR DESCRIPTION
**Pull request by cyclinder**
Signed-off-by: cyclinder <qifeng.guo@daocloud.io>

#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
fix to prevent concurrent map read and map write in kube-proxy
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/103260

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/proxy/ipvs/graceful_termination.go [hunk 1]
```diff
@@ -90,7 +92,9 @@ func (q *graceTerminateRSList) flushList(handler func(rsToDelete *listItem) (boo
 		}
 		if deleted {
 			klog.InfoS("Removed real server from graceful delete real server list", "realServer", name)
-			q.remove(rs)
+			if _, ok := q.list[rs.String()]; ok {
+				delete(q.list, rs.String())
+			}
```
Comment (aojea) [inline pkg/proxy/ipvs/graceful_termination.go:97]: in golang deleting a non-existing key from a map is a nop, I think we can do 
```suggestion
		delete(q.list, rs.String())
```

Comment (aojea) [reply pkg/proxy/ipvs/graceful_termination.go:97]: https://go.dev/doc/effective_go#maps

File: pkg/proxy/ipvs/graceful_termination_test.go [hunk 2]
```diff
@@ -17,12 +17,14 @@ limitations under the License.
 package ipvs
 
 import (
+	"fmt"
+	netutils "k8s.io/utils/net"
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/graceful_termination_test.go:21]: nit: group this import with other k8s.io imports below

File: pkg/proxy/ipvs/graceful_termination.go [hunk 3]
```diff
@@ -90,7 +92,9 @@ func (q *graceTerminateRSList) flushList(handler func(rsToDelete *listItem) (boo
 		}
 		if deleted {
 			klog.InfoS("Removed real server from graceful delete real server list", "realServer", name)
-			q.remove(rs)
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/graceful_termination.go:93]: Can we just delete this function now or are there other calls to q.remove now?

Comment (cyclinder) [reply pkg/proxy/ipvs/graceful_termination.go:93]: can't delete `q.remove()`, there other still calls to `q.remove()`, for example:
https://github.com/kubernetes/kubernetes/blob/7d9a6d1de6552f66ca53ee69e072f604b4368c85/pkg/proxy/ipvs/graceful_termination.go#L200

File: pkg/proxy/ipvs/graceful_termination.go (hunk 1 above)
Comment (cyclinder) [reply pkg/proxy/ipvs/graceful_termination.go:97]: Thanks, Done

File: pkg/proxy/ipvs/graceful_termination_test.go [hunk 4]
```diff
@@ -400,3 +401,36 @@ func Test_GracefulDeleteRS(t *testing.T) {
...
+	go func() {
+		for i := 1; i <= 10; i++ {
+			for j := 1; i <= 100; j++ {
+				gracefulTerminationManager.rsList.add(makeListItem(i, j))
+			}
+		}
+	}()
+
+	if !gracefulTerminationManager.rsList.flushList(gracefulTerminationManager.deleteRsFunc) {
+		t.Errorf("failed to flush entries")
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/graceful_termination_test.go:418]: nit: t.Error instead of t.Errorf

File: pkg/proxy/ipvs/graceful_termination_test.go [hunk 5]
```diff
@@ -400,3 +401,36 @@ func Test_GracefulDeleteRS(t *testing.T) {
 		})
 	}
 }
+
+func Test_RaceTerminateRSList(t *testing.T) {
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/graceful_termination_test.go:405]: So does running this test without the fix result in a panic?

**Review by andrewsykim — COMMENTED** [review]
/approve

deferring lgtm to @uablrek

File: pkg/proxy/ipvs/graceful_termination_test.go (hunk 5 above)
Comment (cyclinder) [reply pkg/proxy/ipvs/graceful_termination_test.go:405]: Yes, if not fixed, running this test will cause panic.
```shell
cyclinder@CycLinder-3 ~/Desktop/code/kubernetes/pkg/proxy/ipvs$ go test -race -timeout 60s -run ^Test_RaceTerminateRSList$  -v                                         ✹ ✭fix_concurrent_map 
=== RUN   Test_RaceTerminateRSList
--- PASS: Test_RaceTerminateRSList (0.00s)
==================
WARNING: DATA RACE
Write at 0x00c0001a85a0 by goroutine 21:
  runtime.mapassign_faststr()
      /Users/cyclinder/Downloads/goland/go1.17/src/runtime/map_faststr.go:202 +0x0
  k8s.io/kubernetes/pkg/proxy/ipvs.(*graceTerminateRSList).add()
      /Users/cyclinder/Desktop/code/kubernetes/pkg/proxy/ipvs/graceful_termination.go:66 +0x2a4
  k8s.io/kubernetes/pkg/proxy/ipvs.Test_RaceTerminateRSList.func1()
      /Users/cyclinder/Desktop/code/kubernetes/pkg/proxy/ipvs/graceful_termination_test.go:412 +0x4f

```

File: pkg/proxy/ipvs/graceful_termination_test.go (hunk 4 above)
Comment (cyclinder) [reply pkg/proxy/ipvs/graceful_termination_test.go]: Thanks, Done.

**Comment by aojea** [comment]
/test pull-kubernetes-integration

:thinking:  it seems it timed out

**Comment by aojea** [comment]
LGTM too
defer to Lars too

**Comment by cyclinder** [comment]
> /test pull-kubernetes-integration
> 
> 🤔 it seems it timed out

I don't know why it timed out...

**Comment by cyclinder** [comment]
Thanks @aojea @andrewsykim @uablrek

**Comment by aojea** [comment]
@liggitt it seems we are hitting again timeouts on the integration job? did you notice something?

/test pull-kubernetes-integration

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/107765 just started populating openapi in all integration in order to support server side apply. 

Openapi construction is expensive, is that contributing to the timeout?

cc @apelisse @Jefftree

**Comment by MadhavJivrajani** [comment]
Thanks for working on this @cyclinder!

File: pkg/proxy/ipvs/graceful_termination_test.go (hunk 5 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/graceful_termination_test.go:405]: thanks!

**Comment by JornShen** [comment]
@cyclinder  consider cherrying pick to the lower version? cc @aojea

**Comment by aojea** [comment]
if there are users affected by the bug I think that is legit asking for the backport,

**Comment by JornShen** [comment]
> This PR fixed for 1.24, Maybe need also fix for 1.22 and 1.23, I'm ok, but need ask for other. /cc @aojea

+1

**Comment by cyclinder** [comment]
Ok, I'll try to cp it.

**Pull request merged**